packages/migrations/src/batched-migrations/batched-migration.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  loadSqlEquiv,
+  queryAsync,
+  queryValidatedOneRow,
+  queryValidatedRows,
+  queryValidatedZeroOrOneRow,
+} from '@prairielearn/postgres';
+import { z } from 'zod';
+
+const sql = loadSqlEquiv(__filename);
+
+export const BatchedMigrationStatusSchema = z.enum([
+  'pending',
+  'paused',
+  'running',
+  'finalizing',
+  'failed',
+  'succeeded',
+]);
+export type BatchedMigrationStatus = z.infer<typeof BatchedMigrationStatusSchema>;
+
+export const BatchedMigrationRowSchema = z.object({
+  id: z.string(),
+  project: z.string(),
+  filename: z.string(),
+  timestamp: z.string(),
+  batch_size: z.number(),
+  min_value: z.bigint({ coerce: true }),
+  max_value: z.bigint({ coerce: true }),
+  status: BatchedMigrationStatusSchema,
+  created_at: z.date(),
+  updated_at: z.date(),
+  started_at: z.date().nullable(),
+});
+export type BatchedMigrationRow = z.infer<typeof BatchedMigrationRowSchema>;
+
+export interface BatchedMigrationParameters {
+  min?: bigint | null;
+  max: bigint | null;
+  batchSize?: number;
+}
+
+/**
+ * A batched migration operates on a range of IDs in batches. It's designed
+ * for migrations that are too large to run in a single transaction.
+ *
+ * Migrations should be idempotent, as they may be run multiple times in case
+ * of unexpected failure.
+ */
+export abstract class BatchedMigration {
+  public abstract getParameters(): Promise<BatchedMigrationParameters>;
+  public abstract execute(start: bigint, end: bigint): Promise<void>;
+}
```

Comment:
This class doesn't actually do any class-y things. I'm thinking maybe we should "make" migrations like this instead?

```ts
import { makeBatchedMigration } from '@prairielearn/migration';

export default makeBatchedMigration({
  getParameters() {
    return {}
  },
  async execute() {
    // ...
  },
});
```

Not having to export something new-able gives us slightly more flexibility in the future. (This is probably bikeshedding.)

Reply:
I'm also generally in favor of avoiding classes wherever possible.

Reply:
I'm fine with whatever you prefer here, although my bias is certainly against classes 😄